pkg/queryfrontend/queryinstant_codec.go

```diff
@@ -283,22 +302,65 @@ func vectorMerge(resps []*queryrange.PrometheusInstantQueryResponse) *queryrange
 	if len(output) == 0 {
 		return &queryrange.Vector{
 			Samples: make([]*queryrange.Sample, 0),
-		}
+		}, nil
 	}
 
-	keys := make([]string, 0, len(output))
-	for key := range output {
-		keys = append(keys, key)
+	type pair struct {
+		metric string
+		s      *queryrange.Sample
 	}
-	sort.Strings(keys)
 
+	samples := make([]*pair, 0, len(output))
+	for k, v := range output {
+		samples = append(samples, &pair{
+			metric: k,
+			s:      v,
+		})
+	}
+
+	sort.Slice(samples, func(i, j int) bool {
+		// Order is determined by the sortFn in the query.
+		if sortAsc {
+			return samples[i].s.Sample.Value < samples[j].s.Sample.Value
+		} else if sortDesc {
+			return samples[i].s.Sample.Value > samples[j].s.Sample.Value
+		} else {
+			// Fallback on sorting by labels.
+			return samples[i].metric < samples[j].metric
+		}
+	})
 	result := &queryrange.Vector{
 		Samples: make([]*queryrange.Sample, 0, len(output)),
 	}
-	for _, key := range keys {
-		result.Samples = append(result.Samples, output[key])
+	for _, p := range samples {
+		result.Samples = append(result.Samples, p.s)
 	}
-	return result
+	return result, nil
+}
+
+func parseQueryForSort(q string) (bool, bool, error) {
+	expr, err := promqlparser.ParseExpr(q)
+	if err != nil {
+		return false, false, err
+	}
+	var sortAsc bool = false
+	var sortDesc bool = false
+	promqlparser.Inspect(expr, func(n promqlparser.Node, _ []promqlparser.Node) error {
+		if n, ok := n.(*promqlparser.Call); ok {
+			if n.Func != nil {
+				if n.Func.Name == "sort" {
+					sortAsc = true
+					return errors.New("done")
```

Comment:
Can we use a global variable for such kind of error? So that we don't have to allocate the struct every time.